subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecHandle.java

```diff
@@ -231,7 +232,7 @@ public ExecHandle start() {
             setState(ExecHandleState.STARTING);
 
             execHandleRunner = new ExecHandleRunner(this, streamsHandler, processLauncher, executorFactory);
-            executor.execute(execHandleRunner);
+            executor.execute(new BuildOperationIdentifierPreservingRunnable(execHandleRunner));
```

Comment:
OK... so I think the way this is working is we capture the build operation so that it can be restored when we run `execHandleRunner`.   And then in `execHandleRunner`, through the `streamsHandler`, we have more runnables that need to have the current build operation.  Since we use the `BuildOperationIdentifierPreservingRunnable` here, they can get it later.

Could we just capture the build operation when we create the `StreamsForwarder` instead of using `BuildOperationIdentifierPreservingRunnable` to pass it around?

And if not when we create the `StreamsForwarder`, when we call `connectStreams()` since that's when we start the process?  In that case, we'd need to pass the build operation down into the `ExecHandleRunner`.